src/effects/backends/builtin/beatgrindeffect.cpp

```diff
@@ -0,0 +1,196 @@
+    // The minimum of the parameter is zero so the exact center of the knob is 1 beat.
+
+    double send_current = m_pSendParameter->value();
+    double feedback_current = m_pFeedbackParameter->value();
+    double period = m_pDelayParameter->value();
+
+    if (enableState == EffectEnableState::Enabling) {
+        gs.write_position = 0;
+        gs.read_position = 0;
+    }
+
+    int delay_frames;
+    if (groupFeatures.has_beat_length_sec) {
+        // period is a number of beats
+        if (m_pQuantizeParameter->toBool()) {
+            period = std::max(roundToFraction(period, 4), 1/8.0);
+            if (m_pTripletParameter->toBool()) {
+                period /= 3.0;
+            }
+        } else if (period < 1/8.0) {
+            period = 1/8.0;
+        }
+        delay_frames = period * groupFeatures.beat_length_sec * bufferParameters.sampleRate();
+    } else {
+        // period is a number of seconds
+        period = std::max(period, 1/8.0);
+        delay_frames = period * bufferParameters.sampleRate();
+    }
+    VERIFY_OR_DEBUG_ASSERT(delay_frames > 0) {
+        delay_frames = 1;
+    }
+
+    int delay_samples = delay_frames * bufferParameters.channelCount();
+    VERIFY_OR_DEBUG_ASSERT(delay_samples <= gs.delay_buf.size()) {
+        delay_samples = gs.delay_buf.size();
+    }
+
+    RampingValue<CSAMPLE_GAIN> send(send_current, gs.prev_send,
+                                    bufferParameters.framesPerBuffer());
+    // Feedback the delay buffer and then add the new input.
+
+    RampingValue<CSAMPLE_GAIN> feedback(feedback_current, gs.prev_feedback,
+                                        bufferParameters.framesPerBuffer());
+
+    qDebug() << "loop samples" << delay_samples;
+    qDebug() << "write pos" << gs.write_position;
+    qDebug() << "read pos" << gs.read_position;
+    
+    for (unsigned int i = 0;
+            i < bufferParameters.samplesPerBuffer();
+            i += bufferParameters.channelCount()) {
+        CSAMPLE_GAIN send_ramped = send.getNext();
+        CSAMPLE_GAIN feedback_ramped = feedback.getNext();
+        auto bufferedSample = 
+            std::make_unique<CSAMPLE []>(bufferParameters.channelCount());
+
+        for (int channel = 0; channel < bufferParameters.channelCount(); channel++) {
+            if (gs.isRecording) {
+                if (gs.write_position >= gs.delay_buf.size()) {
+                    gs.isRecording = false;
+                }
+                else {
+                    gs.delay_buf[gs.write_position++] = pInput[i + channel];
+                }
+            }
+            if (gs.write_position >= delay_samples) {
+                if(gs.read_position >= delay_samples) {
+                    gs.read_position = 0;
+                }
+                    bufferedSample[channel] = gs.delay_buf[gs.read_position++];
+            }
+            pOutput[i + channel] = SampleUtil::clampSample(
+                pInput[i + channel] * send_ramped +
```

Comment:
The manifest declares this as an addDryToWet effect, so EngineEffectChain adds the dry signal. I'm not sure if this should be an addDryToWet effect or not, but I don't think so.

Reply:
If this isn't an addDryToWet effect, then the effect needs to pass through the dry signal until it starts outputting the wet, otherwise there is a gap of silence until the buffer is recorded.

Reply:
Can you elaborate more on this or point me to the relevant documentation, I am not sure what you mean here. I am still a bit lost on the effects architecture and was indeed wondering why there no dry/wet knobs in any of them.
I do know though that's this should not use a send, I was thinking of changing it to a depth parameter

Reply:
addDryToWet is explained in a comment in EngineEffectChain::process. It is for effects that are added on top of the dry signal (with the dry signal at unity gain). This allows for treating them specially in Dry+Wet mode chains, so that a Reverb after an Echo only effects the echoed signal, not the echo mixed with the input of the echo. Play with the Echoverb HP chain preset for example.

Reply:
I think this effect does need its own dry/wet knob. In Traktor, there is one knob confusingly called Gate which combines a dry/wet knob on the left half with a gater on the right half. I think Traktor hacks these together into one knob because Traktor has no metaknob and has a fixed limit of 3 knob parameters per effect. With Mixxx, we could provide a similar behavior with the metaknob by linking the dry/wet parameter to the left side of the metaknob and the gater parameter to the right side. I'm not sure we should make that the default, but that's how it could imitate Traktor. I'm curious to play with what could be done by having those as separate parameters. The gater parameter isn't strictly necessary for a minimum viable product, but it is a cool feature.

Reply:
> wondering why there no dry/wet knobs in any of them

A dry/wet for each effect is unnecessary because effects have a parameter state at which the effect makes no sound. The chain has a mix knob which controls the level of the whole chain with the behavior of that depending on the mix mode (Dry/Wet or Dry+Wet).

However, I think this effect does need its own dry/wet knob because I can't think of any other way to use a knob to make this effect inaudible and fade it in gradually.

Reply:
Yes, +1 for the dry/wet knob for the effect. But I think this should be a an addDryToWet effect too. It makes sense to add an echo and a reverb just the to recorded loop and it makes sense too, to be able to listen to only the recorded loop or just a little touch of it. 
Still not sure how the effect chain adds the dry signal though so I will leave this for later

Reply:
> add an echo and a reverb just the to recorded loop

This can be done by putting the chain in Dry+Wet mode if this effect doesn't pass through the dry signal and isn't an addDryToWet effect.

> listen to only the recorded loop

This requires that the effect is not an addDryToWet effect.

Reply:
So it should not be addDryToWet effect

Reply:
Yeah, after playing with it, I'm pretty sure this effect shouldn't be addDryToWet.